### PR TITLE
MTOM/DR Improvements

### PR DESCRIPTION
**Pull request by jonahkaye**
Ticket: #[1667](https://github.com/metriport/metriport-internal/issues/1667)

### Description

- edge case + unit tests where MTOM doesn't actually have a mutlipart message and still has document raw inside
- redox error where they dont accept more than 1 DocumentReference requests per DR
- removing captures for common xcpd errors flooding sentry
- adding more descriptive logging when we get errors

### Testing

- Local
  - [x] pre-prod testing setup 
  - [x] added many more unit tests

Check each PR.

### Release Plan

- [ ] Merge this

### CODE REVIEW

File: packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/create/iti39-envelope.ts [hunk 1]
```diff
@@ -59,7 +60,7 @@ export function createITI39SoapEnvelope({
       "urn:RetrieveDocumentSetRequest": {
         "urn:DocumentRequest": documentReferences.map(docRef => ({
           "urn:HomeCommunityId": wrapIdInUrnOid(docRef.homeCommunityId),
-          "urn:RepositoryUniqueId": wrapIdInUrnOid(docRef.repositoryUniqueId),
+          "urn:RepositoryUniqueId": docRef.repositoryUniqueId,
```
Comment (jonahkaye) [inline packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/create/iti39-envelope.ts:63]: remove this - surescripts doesnt like and doesnt break others

File: packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/create/iti39-envelope.ts [hunk 2]
```diff
@@ -111,8 +112,19 @@ export function createAndSignBulkDRRequests({
   const signedRequests: BulkSignedDR[] = [];
 
   for (const bodyData of bulkBodyData) {
-    const signedRequest = createAndSignDRRequest(bodyData, samlCertsAndKeys);
-    signedRequests.push({ gateway: bodyData.gateway, signedRequest, outboundRequest: bodyData });
+    const documentReferences = bodyData.documentReference;
+    for (let i = 0; i < documentReferences.length; i += maxDocumentReferencesPerDrRequest) {
```
Comment (jonahkaye) [inline packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/create/iti39-envelope.ts:116]: many epic endpoints fail if we ask for more than 10 documents. This happens in prod often. Batching DR requests into 10 document refrerence chunks

File: packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/mtom/parser.ts [hunk 3]
```diff
@@ -113,22 +119,27 @@ export function parseMTOMResponse(mtomMessage: string, contentType: string): Doc
         removeNSPrefix: true,
       });
       const jsonObj = parser.parse(content);
+      console.log("jsonObj", JSON.stringify(jsonObj, null, 2));
```
Comment (jonahkaye) [inline packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/mtom/parser.ts:122]: nit: leftover

File: packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/mtom/parser.ts [hunk 4]
```diff
@@ -113,22 +119,27 @@ export function parseMTOMResponse(mtomMessage: string, contentType: string): Doc
...
-        // temporarily skip non-xml documents
-        if (
-          docResponse.mimeType === XML_APP_MIME_TYPE ||
-          docResponse.mimeType === XML_TXT_MIME_TYPE
-        ) {
-          documentResponses.push({
-            ...docResponse,
-            Document: stripCidPrefix(docResponse.Document.Include._href),
-          });
+        if (containsMultipartCidReference(docResponse)) {
```
Comment (jonahkaye) [inline packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/mtom/parser.ts:130]: adding handling for the case where response is MTOM but document is still in SOAP. This is how athena sends

File: packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/process/dq-response.ts [hunk 5]
```diff
@@ -117,7 +119,7 @@ function parseDocumentReference(
   }
 
   const documentReference: DocumentReference = {
-    homeCommunityId: outboundRequest.gateway.homeCommunityId,
+    homeCommunityId: stripUrnPrefix(extrinsicObject._home),
```
Comment (jonahkaye) [inline packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/process/dq-response.ts:122]: reverting to this- surescripts depends on gateway they return to us

File: packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/create/iti39-envelope.ts (hunk 2 above)
Comment (jonahkaye) [reply packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/create/iti39-envelope.ts:116]: redox endpoints are failing if we send them more than one DR at a time

File: packages/core/src/external/carequality/ihe-gateway-v2/saml/saml-client.ts [hunk 6]
```diff
@@ -10,6 +10,7 @@ import { creatMtomContentTypeAndPayload } from "../outbound/xca/mtom/builder";
 const { log } = out("Saml Client");
 
 const timeout = 120000;
+const rejectUnauthorized = Config.isDev() ? false : true;
```
Comment (jonahkaye) [inline packages/core/src/external/carequality/ihe-gateway-v2/saml/saml-client.ts:13]: change to make developing in dev easier

**Review by Goncharo — APPROVED** [review]
approved pending nits

File: packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/create/iti39-envelope.ts [hunk 7]
```diff
@@ -10,6 +10,14 @@ import { OutboundDocumentRetrievalReq, XCAGateway } from "@metriport/ihe-gateway
 
 const action = "urn:ihe:iti:2007:CrossGatewayRetrieve";
 
+const onlyOneDocumentReferencePerRequestList = [
+  "2.16.840.1.113883.3.6448",
+  "2.16.840.1.113883.3.6147.458",
+  "2.16.840.1.113883.3.6147.458.2",
```
Comment (Goncharo) [inline packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/create/iti39-envelope.ts:16]: nit: would be useful to put these in consts so we can refer to them later as needed - ie
```
const redoxGatewayOid = "2.16.840.1.113883.3.6448";
const onlyOneDocumentReferencePerRequestList = [
  redoxGatewayOid,
  ...
```

nit: `onlyOneDocumentReferencePerRequestList` isn't very semantic - hard to understand what it does at a glance

perhaps `gatewaysThatAcceptOneDocRefPerRequest` would be better?

**Review by Goncharo — COMMENTED** [review]
approved pending nits

File: packages/core/src/external/carequality/ihe-gateway-v2/saml/saml-client.ts (hunk 6 above)
Comment (leite08) [reply packages/core/src/external/carequality/ihe-gateway-v2/saml/saml-client.ts:13]: I think this is too critical to have an alternative flow. All it takes is a bug w/ `isDev` to break an important part of certificate validation.

Isn't there another option to make it `false` while testing local that doesn't rely on something like this?

File: packages/utils/src/saml/pre-prod-tester.ts [hunk 8]
```diff
@@ -1,6 +1,6 @@
 import * as dotenv from "dotenv";
 dotenv.config();
-// keep that ^ on top
```
Comment (leite08) [inline packages/utils/src/saml/pre-prod-tester.ts:3]: Please don't remove that. Try to auto organize imports, you'll see why.

File: packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/create/iti39-envelope.ts [hunk 9]
```diff
@@ -111,8 +119,24 @@ export function createAndSignBulkDRRequests({
...
+      bodyData.gateway.homeCommunityId
+    )
+      ? minDocumentReferencesPerDrRequest
+      : maxDocumentReferencesPerDrRequest;
+    const documentReferences = bodyData.documentReference;
+    for (let i = 0; i < documentReferences.length; i += documentReferencesPerRequest) {
+      const chunkedBodyData = {
+        ...bodyData,
+        documentReference: documentReferences.slice(i, i + documentReferencesPerRequest),
+      };
```
Comment (leite08) [inline packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/create/iti39-envelope.ts:132]: nit: could use Lodash's `chunk` to simplify this code:

```ts
    const chunks = chunk(documentReferences, documentReferencesPerRequest);
    chunks.forEach(docRefs => {
      const chunkedBodyData = {
        ...bodyData,
        documentReference: docRefs,
      };
```

File: packages/core/src/external/carequality/ihe-gateway-v2/saml/saml-client.ts (hunk 6 above)
Comment (jonahkaye) [reply packages/core/src/external/carequality/ihe-gateway-v2/saml/saml-client.ts]: I used the new pattern you taught me with overriding the variables instead

File: packages/core/src/external/carequality/ihe-gateway-v2/outbound/xcpd/send/xcpd-requests.ts [hunk 10]
```diff
@@ -45,23 +43,17 @@ export async function sendSignedXCPDRequests({
...
-    } catch (error) {
+      //eslint-disable-next-line @typescript-eslint/no-explicit-any
+    } catch (error: any) {
       const msg = "HTTP/SSL Failure Sending Signed XCPD SAML Request";
-      log(`${msg}, error: ${error}`);
+      log(
+        `${msg}, cxId: ${cxId}, patientId: ${patientId}, gateway: ${request.gateway.oid}, error: ${error}`
+      );
+      if (error?.response?.data) {
+        log(`error details: ${error?.response?.data}`);
```
Comment (jonahkaye) [inline packages/core/src/external/carequality/ihe-gateway-v2/outbound/xcpd/send/xcpd-requests.ts:53]: important information for debugging we are missing by not logging this

File: packages/core/src/external/carequality/ihe-gateway-v2/outbound/xcpd/process/xcpd-response.ts [hunk 11]
```diff
@@ -116,17 +116,6 @@ function handlePatientErrorResponse({
   patientId?: string | undefined;
   cxId?: string | undefined;
 }): OutboundPatientDiscoveryResp {
-  const msg = "An AbortedError (AE) was received from the responding gateway";
```
Comment (jonahkaye) [inline packages/core/src/external/carequality/ihe-gateway-v2/outbound/xcpd/process/xcpd-response.ts:119]: stop flooding sentry

File: packages/core/src/external/carequality/ihe-gateway-v2/outbound/xcpd/send/xcpd-requests.ts [hunk 12]
```diff
@@ -45,23 +43,17 @@ export async function sendSignedXCPDRequests({
...
+      }
 
       const errorString: string = errorToString(error);
-      const extra = {
-        errorString,
-        request,
-        patientId,
-        cxId,
-      };
-      capture.error(msg, {
```
Comment (jonahkaye) [inline packages/core/src/external/carequality/ihe-gateway-v2/outbound/xcpd/send/xcpd-requests.ts:59]: stop flooding sentry

File: packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/send/dq-requests.ts [hunk 13]
```diff
@@ -45,10 +45,15 @@ export async function sendSignedDQRequests({
...
+      //eslint-disable-next-line @typescript-eslint/no-explicit-any
+    } catch (error: any) {
       const msg = "HTTP/SSL Failure Sending Signed DQ SAML Request";
-      log(`${msg}, error: ${error}`);
-
+      log(
+        `${msg}, cxId: ${cxId}, patientId: ${patientId}, gateway: ${request.gateway.homeCommunityId}, error: ${error}`
+      );
+      if (error?.response?.data) {
+        log(`error details: ${error?.response?.data}`);
```
Comment (leite08) [inline packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/send/dq-requests.ts:55]: This will likely log `[object Object]`

File: packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/send/dr-requests.ts [hunk 14]
```diff
@@ -51,9 +51,15 @@ export async function sendSignedDRRequests({
...
-    } catch (error) {
+      //eslint-disable-next-line @typescript-eslint/no-explicit-any
+    } catch (error: any) {
       const msg = "HTTP/SSL Failure Sending Signed DR SAML Request";
-      log(`${msg}, error: ${error}`);
+      log(
+        `${msg}, cxId: ${cxId}, patientId: ${patientId}, gateway: ${request.gateway.homeCommunityId}, error: ${error}`
+      );
+      if (error?.response?.data) {
+        log(`error details: ${error?.response?.data}`);
```
Comment (leite08) [inline packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/send/dr-requests.ts:61]: ditto
